File: backend/src/Anela.Heblo.Persistence/Features/Authorization/GroupClosure.cs
using Anela.Heblo.Domain.Features.Authorization.Entities;

namespace Anela.Heblo.Persistence.Features.Authorization;

/// <summary>Cycle-safe transitive closure of group nesting → union of permissions.</summary>
public static class GroupClosure
{
    public static IReadOnlyCollection<string> Resolve(
        IEnumerable<Guid> directGroupIds,
        IReadOnlyCollection<GroupPermission> allPermissions,
        IReadOnlyCollection<GroupParent> allParents)
    {
        var permsByGroup = allPermissions
            .GroupBy(p => p.GroupId)
            .ToDictionary(g => g.Key, g => g.Select(p => p.PermissionValue).ToArray());
        var parentsByGroup = allParents
            .GroupBy(p => p.GroupId)
            .ToDictionary(g => g.Key, g => g.Select(p => p.ParentGroupId).ToArray());

        var visited = new HashSet<Guid>();
        var queue = new Queue<Guid>(directGroupIds);
        var result = new HashSet<string>(StringComparer.Ordinal);

        while (queue.Count > 0)
        {
            var groupId = queue.Dequeue();
            if (!visited.Add(groupId)) continue; // already processed → cycle/diamond safe

            if (permsByGroup.TryGetValue(groupId, out var perms))
                foreach (var p in perms) result.Add(p);

            if (parentsByGroup.TryGetValue(groupId, out var parents))
                foreach (var parent in parents) queue.Enqueue(parent);
        }

        return result;
    }
}

File: frontend/src/components/access-management/groupClosure.ts
export type GroupNode = {
  id: string;
  permissions: string[];
  parentGroupIds: string[];
};

/**
 * Cycle-safe transitive closure of group nesting → set of inherited permissions.
 *
 * TS port of backend GroupClosure.Resolve. Seeds the traversal with the selected
 * parent (included) group IDs only, so the result contains permissions inherited
 * from nested groups — never the editing group's own direct permissions.
 */
export function resolveInheritedPermissions(
  selectedParentIds: string[],
  groups: GroupNode[],
): Set<string> {
  const groupsById = new Map(groups.map((g) => [g.id, g]));

  const visited = new Set<string>();
  const queue: string[] = [...selectedParentIds];
  const result = new Set<string>();

  while (queue.length > 0) {
    const groupId = queue.shift()!;
    if (visited.has(groupId)) continue; // already processed → cycle/diamond safe
    visited.add(groupId);

    const group = groupsById.get(groupId);
    if (!group) continue;

    for (const permission of group.permissions) result.add(permission);
    for (const parentId of group.parentGroupIds) queue.push(parentId);
  }

  return result;
}
